File: lib/devouch/policy.rb
# ABOUTME: Applies each repository's explicit acceptance policy to independently valid evidence.
# ABOUTME: Never lets a credential select its own policy or weaken an unsupported issuer threshold.
require "digest"
require_relative "files"

module Devouch
  class Policy
    ADDRESS = /\A0x[0-9a-fA-F]{40}\z/
    attr_reader :data, :digest

    def initialize(raw)
      @data = Files.parse(raw)
      @digest = "sha256:#{Digest::SHA256.hexdigest(raw)}"
      valid = @data.is_a?(Hash) &&
        @data.keys.sort == %w[repositoryId chainId trustedIssuers allowedScopes allowedResolvers requiredIssuers].sort &&
        @data["repositoryId"].is_a?(String) && @data["repositoryId"].match?(/\A[a-zA-Z0-9_.-]+\/[a-zA-Z0-9_.-]+\z/) &&
        @data["chainId"] == 11155111 && @data["requiredIssuers"] == 1 &&
        array?("trustedIssuers") { |value| address?(value) } &&
        array?("allowedScopes") { |value| value.is_a?(String) && value.match?(/\A[a-z][a-z0-9-]{0,63}\z/) } &&
        array?("allowedResolvers") { |value| value.is_a?(Hash) && value.keys.sort == %w[address implementation] &&
          address?(value["address"]) && address?(value["implementation"]) }
      raise JSON::ParserError unless valid
    rescue JSON::ParserError
      raise Error.new("invalid_policy", "Use a Sepolia policy with requiredIssuers: 1 and explicit issuer, scope, and resolver lists.")
    end

    def description
      {"repository_id" => @data.fetch("repositoryId"), "digest" => @digest}
    end

    def reasons(evidence)
      reasons = []
      reasons << "issuer_not_trusted" unless @data["trustedIssuers"].any? { |issuer| equal_address?(issuer, evidence["issuer"]) }
      reasons << "scope_not_allowed" unless @data["allowedScopes"].include?(evidence["scope"])
      reasons << "resolver_not_allowed" unless @data["allowedResolvers"].any? { |entry|
        equal_address?(entry["address"], evidence["resolver"]) && equal_address?(entry["implementation"], evidence["implementation"])
      }
      reasons
    end

    private

    def equal_address?(left, right)
      left.downcase == right.to_s.downcase
    end

    def address?(value)
      value.is_a?(String) && value.match?(ADDRESS) && value.downcase != "0x#{"0" * 40}"
    end

    def array?(key, &predicate)
      value = @data[key]
      value.is_a?(Array) && value.length <= 64 && value.all?(&predicate)
    end
  end
end

File: src/policy.ts
// ABOUTME: Applies the same explicit repository policy contract as the Ruby CLI.
// ABOUTME: Evaluates only valid evidence and preserves the CLI reason order for shared fixture tests.
import { exactKeys, object, strictJson } from './credential';
import { EvidenceError, insist, type EvidenceStatus } from './errors';

export type RepositoryPolicy = { repositoryId: string; chainId: number; trustedIssuers: string[];
  allowedScopes: string[]; allowedResolvers: { address: string; implementation: string }[]; requiredIssuers: number };
export type PolicyEvidence = { evidence_status: EvidenceStatus; reason_codes: string[];
  issuer?: string | null; scope?: string | null; resolver?: string | null; implementation?: string | null };
export type PolicyDecision = { policy_status: 'accepted' | 'rejected' | 'not_evaluated'; reason_codes: string[] };
const isAddress = (value: unknown): value is string => typeof value === 'string'
  && /^0x[0-9a-fA-F]{40}(?![\s\S])/.test(value) && !/^0x0{40}$/.test(value);
const list = (value: unknown, predicate: (item: unknown) => boolean) =>
  Array.isArray(value) && value.length <= 64 && value.every(predicate);
const sameAddress = (left: string, right: string | null | undefined) => left.toLowerCase() === (right ?? '').toLowerCase();

export function parsePolicy(raw: string): RepositoryPolicy {
  try {
    const p = object(strictJson(raw, 16384));
    exactKeys(p, ['repositoryId', 'chainId', 'trustedIssuers', 'allowedScopes', 'allowedResolvers', 'requiredIssuers']);
    insist(typeof p.repositoryId === 'string' && /^[a-zA-Z0-9_.-]+\/[a-zA-Z0-9_.-]+(?![\s\S])/.test(p.repositoryId));
    insist(p.chainId === 11155111 && p.requiredIssuers === 1);
    insist(list(p.trustedIssuers, isAddress));
    insist(list(p.allowedScopes, value => typeof value === 'string' && /^[a-z][a-z0-9-]{0,63}(?![\s\S])/.test(value)));
    insist(list(p.allowedResolvers, value => {
      const entry = object(value); exactKeys(entry, ['address', 'implementation']);
      return isAddress(entry.address) && isAddress(entry.implementation);
    }));
    return p as RepositoryPolicy;
  } catch { throw new EvidenceError('invalid_policy'); }
}

export function evaluatePolicy(policy: RepositoryPolicy, evidence: PolicyEvidence): PolicyDecision {
  if (evidence.evidence_status !== 'valid') return { policy_status: 'not_evaluated', reason_codes: [...evidence.reason_codes] };
  const reasons: string[] = [];
  if (!policy.trustedIssuers.some(issuer => sameAddress(issuer, evidence.issuer))) reasons.push('issuer_not_trusted');
  if (!policy.allowedScopes.includes(evidence.scope ?? '')) reasons.push('scope_not_allowed');
  if (!policy.allowedResolvers.some(entry => sameAddress(entry.address, evidence.resolver)
    && sameAddress(entry.implementation, evidence.implementation))) reasons.push('resolver_not_allowed');
  return { policy_status: reasons.length ? 'rejected' : 'accepted', reason_codes: reasons };
}
